src/Scene/Components/ModelComponent.cc

```diff
@@ -0,0 +1,226 @@
+#include "ModelComponent.hh"
+
+// libs
+#define TINYOBJLOADER_IMPLEMENTATION
+#include "external/tinyobjloader/tiny_obj_loader.h"
+#include "external/glm/glm/gtx/hash.hpp"
+
+// std
+#include <cassert>
+#include <unordered_map>
+
+#ifndef ENGINE_DIR
+#define ENGINE_DIR "../"
+#endif
+
+using namespace esp;
+
+namespace std
+{
+	template<>
+	struct hash<esp::ModelComponent::Vertex>
+	{
+		size_t operator()(const esp::ModelComponent::Vertex& vertex) const
+		{
+			size_t seed = 0;
+			esp::hashCombine(seed, vertex.m_position, vertex.m_color, vertex.m_normal, vertex.m_uv);
+			return seed;
+		}
+	};
+}
+
+std::vector<VkVertexInputBindingDescription> ModelComponent::Vertex::get_binding_descriptions()
+{
+	std::vector<VkVertexInputBindingDescription> binding_descriptions(1);
+	binding_descriptions[0].binding = 0;
+	binding_descriptions[0].stride = sizeof(Vertex);
+	binding_descriptions[0].inputRate = VK_VERTEX_INPUT_RATE_VERTEX;
+
+	return binding_descriptions;
+}
+
+std::vector<VkVertexInputAttributeDescription> ModelComponent::Vertex::get_attribute_descriptions()
+{
+	std::vector<VkVertexInputAttributeDescription> attribute_descriptions{};
+
+	attribute_descriptions.push_back({ 0, 0, VK_FORMAT_R32G32B32_SFLOAT, offsetof(Vertex, m_position) });
+	attribute_descriptions.push_back({ 1, 0, VK_FORMAT_R32G32B32_SFLOAT, offsetof(Vertex, m_color) });
+	attribute_descriptions.push_back({ 2, 0, VK_FORMAT_R32G32B32_SFLOAT, offsetof(Vertex, m_normal) });
+	attribute_descriptions.push_back({ 3, 0, VK_FORMAT_R32G32_SFLOAT, offsetof(Vertex, m_uv) });
+
+	// Alternative
+	/*attribute_descriptions[0].binding = 0;
+	attribute_descriptions[0].location = 0;
+	attribute_descriptions[0].format = VK_FORMAT_R32G32B32_SFLOAT;
+	attribute_descriptions[0].offset = offsetof(Vertex, position);
+	attribute_descriptions[1].binding = 0;
+	attribute_descriptions[1].location = 1;
+	attribute_descriptions[1].format = VK_FORMAT_R32G32B32_SFLOAT;
+	attribute_descriptions[1].offset = offsetof(Vertex, color);*/
+
+	return attribute_descriptions;
+}
+
+void ModelComponent::Builder::loadModel(const std::string& filepath)
+{
+	tinyobj::attrib_t attrib;
+	std::vector<tinyobj::shape_t> shapes;
+	std::vector<tinyobj::material_t> materials;
+	std::string warn, err;
+
+	std::string engine_path = ENGINE_DIR + filepath;
```

Comment:
Why are model files read relatively to the engines location? I think file paths should be absolute. Also if ENGINE_DIR isn't ended by '/' symbol this line can cause runtime errors.

Reply:
https://github.com/engineer-boys/espert-core/pull/9#discussion_r1360602413